File: julia/state-of-tic-tac-toe/1/state-of-tic-tac-toe.jl
using LinearAlgebra

const x = 1
const o = -1

function wins(b_int, player)
    # It's a win if any of the rows, columns, or diagonals add up to ±3
    row_wins = count(sum(row) == 3 * player for row in eachrow(b_int))
    col_wins = count(sum(col) == 3 * player for col in eachcol(b_int))
    diag_wins = (sum(diag(b_int)) == 3 * player ? 1 : 0) + 
                (sum(diag(rotr90(b_int))) == 3 * player ? 1 : 0)
    (row_wins, col_wins, diag_wins)
end

function valid_double_win(wins)
    rows, cols, diags = wins
    sum(wins) == 2 && rows < 2 && cols < 2
end

function gamestate(board)
    # Parse to Int Matrix
    b_str = hcat(split.(board, "")...)
    replace!(b_str, "X" => "1")
    replace!(b_str, " " => "0")
    replace!(b_str, "O" => "-1")
    b_int = parse.(Int, b_str)

    sum(b_int) ∉ (0, 1) && error("Must take turns in order!")

    x_wins = wins(b_int, x)
    o_wins = wins(b_int, o)
    (sum(x_wins) > 0 && sum(o_wins) > 0) && error("Too many wins!")

    total_wins = sum(x_wins) + sum(o_wins)
    if total_wins == 1 || valid_double_win(x_wins) || valid_double_win(o_wins)
        "win"
    elseif total_wins > 1
        error("Too many wins!")
    elseif sum(b_int) == 1 && !any(x == 0 for x in b_int)
        "draw"
    else
        "ongoing"
    end
end

File: python/state-of-tic-tac-toe/state_of_tic_tac_toe.py
# values to represent each player
PLAYER_X = 1
PLAYER_O = -1


def gamestate(board):
    # convert board to integers, so we can do arithmetic
    values = {'X': PLAYER_X, ' ': 0, 'O': PLAYER_O}
    int_board = [[values[pos] for pos in row] for row in board]

    # preliminary checks for validity
    board_total = sum([sum(row) for row in int_board])
    if board_total > 1:
        raise ValueError("Wrong turn order: X went twice")
    if board_total < 0:
        raise ValueError("Wrong turn order: O started")

    # winning lines for each player
    x_wins = wins(int_board, PLAYER_X)
    o_wins = wins(int_board, PLAYER_O)
    if sum(x_wins) > 0 and sum(o_wins) > 0:
        raise ValueError("Impossible board: game should have ended after the game was won")

    # work out result for a valid board
    total_wins = sum(x_wins) + sum(o_wins)
    if total_wins == 1 or valid_double_win(x_wins) or valid_double_win(o_wins):
        return 'win'
    if any([n == 0 for row in int_board for n in row]):
        return 'ongoing'
    return 'draw'


def valid_double_win(wins):
    # a player can win on both diagonals simultaneously
    rows, cols, diags = wins
    return sum(wins) == 2 and rows < 2 and cols < 2


def wins(int_board, player):
    row_wins = sum([sum(row) == 3 * player for row in int_board])

    transpose_board = list(map(list, zip(*int_board)))
    col_wins = sum([sum(row) == 3 * player for row in transpose_board])

    diag1 = sum([int_board[n][n] for n in (0, 1, 2)])
    diag2 = sum([int_board[n][2 - n] for n in (0, 1, 2)])
    diag_wins = (diag1 == 3 * player) + (diag2 == 3 * player)
    return (row_wins, col_wins, diag_wins)
